src/charm.py

```diff
@@ -53,21 +58,107 @@ def _on_peer_relation_changed(self, event):
+        """
+
+        if not self.unit.is_leader():
+            return
+
+        if not self.mysql.is_ready():
+            msg = "Waiting for MySQL Service"
+            self.unit.status = WaitingStatus(msg)
+            logger.debug(msg)
+            event.defer()
+            return
+
+        self._on_update_status(event)
+        self._stored.mysql_initialized = True
+        self.unit.status = ActiveStatus()
+
+    # Handles update-status event
+    def _on_update_status(self, event):
+        """Set status for all units
+        Status may be
+        - MySQL is not ready,
+        - MySQL is not Initialized
+        - Unit is active
+        """
+        if not self.unit.is_leader():
+            self.unit.status = ActiveStatus()
+            return
+
+        if not self.mysql.is_ready():
+            status_message = "MySQL not ready yet"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        if not self._stored.mysql_initialized:
+            status_message = "MySQL not initialized"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _provide_mysql(self) -> None:
+        if self._stored.mysql_initialized:
+            self.mysql_provider = MySQLProvider(
+                self, "database", self.provides
+            )
+            self.mysql_provider.ready()
+            logger.info("MySQL Provider is available")
+
+    @property
+    def mysql(self) -> MySQL:
+        """Returns MySQL object"""
+        mysql_config = {
+            "app_name": self.model.app.name,
+            "host": self.hostname,
+            "port": self.model.config["port"],
+            "user_name": "root",
+            "mysql_root_password": self._stored.mysql_setup[
+                "MYSQL_ROOT_PASSWORD"
+            ],
+        }
+        return MySQL(mysql_config)
+
+    @property
+    def hostname(self) -> str:
+        """Unit hostname"""
+        unit_id = self.unit.name.split("/")[1]
+        return "{0}-{1}.{0}-endpoints".format(self.model.app.name, unit_id)
+
     @property
-    def mysql_root_password(self) -> str:
+    def provides(self) -> dict:
+        """Provides dictionary"""
+        provides = {
+            "provides": {"mysql": self.mysql.version()},
+        }
+        return provides
+
+    @property
+    def mysql_root_password(self) -> Union[str, None]:
         """
         This property return MYSQL_ROOT_PASSWORD from StoredState.
         If the password isn't in StoredState, generates one.
         """
 
         if not self.unit.is_leader():
-            return
+            return None
 
         if "MYSQL_ROOT_PASSWORD" not in self._stored.mysql_setup:
-            password = "".join(
-                random.choice(ascii_letters + digits) for x in range(20)
-            )
-            self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"] = password
+            self._stored.mysql_setup[
```

Comment:
Minor, stylistic issue : I do not see why all caps is necessary for a dictionary key. If this was a global constant I could see why but this does not seem to be a global or constant.

Reply:
I used all caps just to follow the style in mysql docker image config options: https://hub.docker.com/r/ubuntu/mysql
Anyway I can change it

Reply:
Your call. I appreciate that environment variables are typically all caps, but no reason you need to use the same capitalization in the environment variable and the dictionary key. Again completely your decision. Just my two cents :-).

Reply:
We'll follow this in #41